plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp

```diff
@@ -0,0 +1,208 @@
+         std::vector<producer_plugin::snapshot_schedule_information> sr;
+         _snapshot_db >> sr;
+         // if db read succeeded, clear/load
+         _snapshot_requests.get<by_snapshot_id>().clear();
+         _snapshot_requests.insert(sr.begin(), sr.end());
+      }
+   }
+
+   // add pending snapshot info to inflight snapshot request
+   void add_pending_snapshot_info(const producer_plugin::snapshot_information & si) {
+      auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+      auto  snapshot_req   = snapshot_by_id.find(_inflight_sid);
+      if (snapshot_req != snapshot_by_id.end()) {
+         _snapshot_requests.modify(snapshot_req, [&si](auto& p) { 
+            if (!p.pending_snapshots) {
+               p.pending_snapshots = std::vector<producer_plugin::snapshot_information>();
+            }
+            p.pending_snapshots->emplace_back(si);       
+         });
+      }
+   }
+
+   // snapshot executor
+   void set_create_snapshot_fn(std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> fn) {
+      _create_snapshot = std::move(fn);
+   }
+
+   void execute_snapshot(uint32_t srid) {
+      _inflight_sid = srid;
+      auto next = [srid, this](const std::variant<fc::exception_ptr, producer_plugin::snapshot_information>& result) {
+         if(std::holds_alternative<fc::exception_ptr>(result)) {
+            try {
+               std::get<fc::exception_ptr>(result)->dynamic_rethrow_exception();
+            } catch(const fc::exception& e) {
+               elog( "snapshot creation error: ${details}", ("details",e.to_detail_string()) );
+               appbase::app().quit();
+            } catch(const std::exception& e) {
+               elog( "snapshot creation error: ${details}", ("details",e.what()) );
+               appbase::app().quit();
+            }
+         } else {
+            // success, snapshot finalized
+            auto snapshot_info   = std::get<producer_plugin::snapshot_information>(result);
+            auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+            auto  snapshot_req   = snapshot_by_id.find(srid);
+
+            if (snapshot_req != snapshot_by_id.end()) {               
+               if (auto pending = snapshot_req->pending_snapshots; pending) {                 
+                  auto it = std::remove_if(pending->begin(), pending->end(), [&snapshot_info](const producer_plugin::snapshot_information & s){ return s.head_block_num <=  snapshot_info.head_block_num; });
+                  pending->erase(it, pending->end());
+                  _snapshot_requests.modify(snapshot_req, [&pending](auto& p) { 
+                     p.pending_snapshots = std::move(pending);
+                  });
+               }
+            } else {     
+               // should not happen, but if happens - not a terminal error          
+               elog( "unable to finalize pending snapshot request: ${details}", ("details", srid) );
```

Comment:
After your last change (addition of this line), I have the following error when running the test:

![image](https://user-images.githubusercontent.com/1496907/225733044-993714ec-f7f0-40eb-ba4d-ea5ef5205d80.png)

Reply:
I believe we hit this because the snapshot_request has already been removed from `_snapshot_requests` by the time `add_pending_snapshot_info` is called (even though we save its id in `_inflight_sid` - and I'm not sure why we do that - it is erased from the multi-index in `unschedule_snapshot`).

I suggest you just comment out this `elog` call for now - with a comment as why it is triggered.

Reply:
Thats totally correct, I was too fast adding this trace. It's normal behavior to have snapshot request removed (expired) and therefore any pendings should be gone as well.